### DISCUSSION TITLE
Liquid Exception: Liquid syntax error (line 16): Unknown tag 'include_cached' in /_layouts/default.htm

### DISCUSSION THREAD

**Discussion by RichStone** [mmistakes/minimal-mistakes#2970] [Q&A] [answered]
## Environment

```
rich@MBP-de-Rich:~/code/src/github.com/richstone.github.io$ bundle exec jekyll --version
jekyll 3.7.4
```

- Minimal Mistakes version: from repo
- Ruby gem or remote theme version: ruby 2.3.7p456 (2018-03-28 revision 63024) [universal.x86_64-darwin17]
- Jekyll version: 3.7.4
- Git repository URL: https://github.com/RichStone/richstone.github.io
- Operating system: MacOS

## Expected behavior

```
rich@MBP-de-Rich:~/code/src/github.com/richstone.github.io$ JEKYLL_ENV=production bundle exec jekyll serve --trace
```

should serve locally

instead there comes this error message:

```
rich@MBP-de-Rich:~/code/src/github.com/richstone.github.io$ JEKYLL_ENV=production bundle exec jekyll serve --trace
Configuration file: /Users/rich/code/src/github.com/richstone.github.io/_config.yml
            Source: /Users/rich/code/src/github.com/richstone.github.io
       Destination: /Users/rich/code/src/github.com/richstone.github.io/_site
 Incremental build: disabled. Enable with --incremental
      Generating... 
      Remote Theme: Using theme mmistakes/minimal-mistakes
  Liquid Exception: Liquid syntax error (line 16): Unknown tag 'include_cached' in /_layouts/default.html
bundler: failed to load command: jekyll (/usr/local/bin/jekyll)
Liquid::SyntaxError: Liquid syntax error (line 16): Unknown tag 'include_cached'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/document.rb:23:in `unknown_tag'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/document.rb:11:in `block in parse'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/block_body.rb:33:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/document.rb:10:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/document.rb:5:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/template.rb:132:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/liquid-4.0.0/lib/liquid/template.rb:116:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/liquid_renderer/file.rb:13:in `block in parse'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/liquid_renderer/file.rb:49:in `measure_time'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/liquid_renderer/file.rb:12:in `parse'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/renderer.rb:118:in `render_liquid'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/renderer.rb:188:in `render_layout'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/renderer.rb:155:in `place_in_layouts'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/renderer.rb:85:in `render_document'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/renderer.rb:62:in `run'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:462:in `block (2 levels) in render_docs'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:460:in `each'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:460:in `block in render_docs'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:459:in `each_value'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:459:in `render_docs'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:190:in `render'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/site.rb:73:in `process'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/command.rb:28:in `process_site'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/build.rb:65:in `build'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/build.rb:36:in `process'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/serve.rb:93:in `block in start'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/serve.rb:93:in `each'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/serve.rb:93:in `start'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/lib/jekyll/commands/serve.rb:75:in `block (2 levels) in init_with_program'
  /Library/Ruby/Gems/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `block in execute'
  /Library/Ruby/Gems/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `each'
  /Library/Ruby/Gems/2.3.0/gems/mercenary-0.3.6/lib/mercenary/command.rb:220:in `execute'
  /Library/Ruby/Gems/2.3.0/gems/mercenary-0.3.6/lib/mercenary/program.rb:42:in `go'
  /Library/Ruby/Gems/2.3.0/gems/mercenary-0.3.6/lib/mercenary.rb:19:in `program'
  /Library/Ruby/Gems/2.3.0/gems/jekyll-3.7.4/exe/jekyll:15:in `<top (required)>'
  /usr/local/bin/jekyll:22:in `load'
  /usr/local/bin/jekyll:22:in `<top (required)>'
```

Unfortunately, I don't remember anymore what could have caused this error. I reverted all changes (which were really minor ones) and pulled the code from master. It runs fine on datagoodie.com and on https://richstone.github.io.

I did a `bundle update` which did not help and also `sudo gem update --system` / `sudo gem update`

Does it have something to do with jekyll / ruby installation or with the theme?

**Comment by mmistakes** [comment, marked as answer]
You haven't done anything wrong. The theme now uses the [jekyll-include-cache](https://github.com/benbalter/jekyll-include-cache) plugin. And while it is supported by GitHub Pages, they don't enabled it by default. You need to opt in and add it to the `plugins` array of your [`_config.yml`](https://github.com/mmistakes/minimal-mistakes/blob/master/_config.yml#L223-L229).

```yaml
plugins:
  - jekyll-paginate
  - jekyll-sitemap
  - jekyll-gist
  - jekyll-feed
  - jemoji
  - jekyll-include-cache
```

**Comment by RichStone** [comment]
Thanks for the fix! 

How do I know in future that the theme was updated with new gems and that I have to install them and add them to config.yaml/gemfile?

**Comment by mmistakes** [comment]
@RichStone check the [CHANGELOG](https://github.com/mmistakes/minimal-mistakes/blob/master/CHANGELOG.md). Any updates made to the theme are noted there.

What I would suggest to avoid future disruptions is you lock the theme to current released version, and upgrade when you want to. This will avoid you pulling down features that may not be backwards compatible with your content.

If you're using the theme as a Ruby gem you'd add the version to your `Gemfile` like this:

```ruby
gem "minimal-mistakes-jekyll", "4.13.0"
```

And if you're using it as a remote theme you can lock it like this in your `_config.yml`:

```yaml
remote_theme: "mmistakes/minimal-mistakes@4.13.0"

**Comment by RichStone** [comment]
Thank you very much! Everything seems to work now, although I was in a nokogiri hell while installing everything on my mac 🙈

**Comment by bjwo** [comment]
I am currently also getting the same error, even though I made sure that `jekyll-include-cache` is listed in the `plugins` section. I tried the a) gem-based method, b) Remote theme method c) cloning the repo. I installed all dependencies using bundler, and `jekyll-include-cache` also gets installed.

I am using the latest release v4.14.2

Any clues what I am doing wrong?

```
bundle install --force
Fetching gem metadata from https://rubygems.org/.........
Installing concurrent-ruby 1.1.4
Installing i18n 0.9.5
Installing minitest 5.11.3
Installing thread_safe 0.3.6
Installing tzinfo 1.2.5
Installing activesupport 5.2.2
Installing public_suffix 3.0.3
Installing addressable 2.5.2
Using bundler 1.17.3
Installing colorator 1.1.0
Installing eventmachine 1.2.7 (x64-mingw32)
Installing http_parser.rb 0.6.0 with native extensions
Installing em-websocket 0.5.1
Installing multipart-post 2.0.0
Installing faraday 0.15.4
Installing ffi 1.10.0 (x64-mingw32)
Installing forwardable-extended 2.6.0
Installing gemoji 3.0.0
Installing mini_portile2 2.4.0
Installing nokogiri 1.10.1 (x64-mingw32)
Installing html-pipeline 2.10.0
Installing rb-fsevent 0.10.3
Installing rb-inotify 0.10.0
Installing sass-listen 4.0.0
Installing sass 3.7.3
Installing jekyll-sass-converter 1.5.2
Installing ruby_dep 1.5.0
Installing listen 3.1.5
Installing jekyll-watch 2.1.2
Installing kramdown 1.17.0
Installing liquid 4.0.1
Installing mercenary 0.3.6
Installing pathutil 0.16.2
Installing rouge 3.3.0
Installing safe_yaml 1.0.4
Installing jekyll 3.8.5
Installing jekyll-data 1.0.0
Installing jekyll-feed 0.11.0
Installing sawyer 0.8.1
Installing octokit 4.13.0
Installing jekyll-gist 1.5.0
Installing jekyll-include-cache 0.1.0
Installing jekyll-paginate 1.1.0
Installing jekyll-sitemap 1.2.0
Installing jemoji 0.10.1
Installing minimal-mistakes-jekyll 4.14.2
Bundle complete! 2 Gemfile dependencies, 46 gems now installed.
```

**Comment by mmistakes** [comment]
@bjwo And you're running Jekyll with `bundle exec jekyll build` correct?

Do you have a public repo I can look at? If you have the include cache plugin in your `Gemfile` and it's under `plugins` in _config.yml then it should work.

**Comment by bjwo** [comment]
@mmistakes Yes, that's correct, and I have the `jekyll-include-cache` in my Gemfile and under plugins.

However, I doubt that it's the repo, but rather my setup, as I also have tried to clone your repo and `bundle exec jekyll serve` that locally after running `bundle install` and that also fails.

Is there any better way how I can debug this? Thanks, Björn

**Comment by mmistakes** [comment]
@bjwo If you have a public repo I can try and build your site and see if I get the same error. If I don't then it's definitely your Ruby environment.

You'd be surprised how often it's something silly that I can easily spot if I can see your actual code in the repo... that's why I asked.

**Comment by sagnikb** [comment]
Hi! I did all of the suggested things - added `jekyll-include-cache` to `plugins`, and tried `bundle install --force` which worked too, but I still get the same error. Any ideas?

@bjwo - did you solve the problem?

**Comment by enjaku** [comment]
A bit new at this but having the same error.
     The tag include_cached on line 16 in /_layouts/default.html is not a recognized Liquid tag...

From config.yml
'#' Plugins (previously gems:)
plugins:
  - jekyll-paginate
  - jekyll-sitemap
  - jekyll-gist
  - jekyll-feed
  - jemoji
  - jekyll-include-cache

'#' mimic GitHub Pages with --safe
whitelist:
  - jekyll-paginate
  - jekyll-sitemap
  - jekyll-gist
  - jekyll-feed
  - jemoji
  - jekyll-include-cache

Gemfile:
source "https://rubygems.org"
gem "jekyll-include-cache"

*** LOCAL GEMS ***
...
jekyll (3.8.6)
jekyll-include-cache (0.2.0)
...

  enjaku.github.io git:(master) ✗ bundle exec jekyll build
           WARNING: Error reading configuration. Using defaults (and options).
Configuration file: (INVALID) /Users/ericstewart/Desktop/enjaku.github.io/_config.yml
            Source: /Users/ericstewart/Desktop/enjaku.github.io
       Destination: /Users/ericstewart/Desktop/enjaku.github.io/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
  Liquid Exception: Liquid syntax error (line 16): Unknown tag 'include_cached' in /_layouts/default.html
jekyll 3.8.6 | Error:  Liquid syntax error (line 16): Unknown tag 'include_cached'

At a bit of a loss here.  I've been playing with this for three days now.

**Comment by hyunnna** [comment]
please help me..
The tag img on line 3 in /_layouts/post.html is not a recognized Liquid tag.

I have this error every I try.

This is my /_layouts/post.html code.

---
layout: base
---
{% if page.cover %}
<div class="row">
{% img page.cover cover %}
</div>
{% endif %}
<div class="row">
<h2>{{ page.title }}</h2>
<div class="valign-wrapper">
<div class="author"><img src="{{ site.author_img | prepend: site.baseurl }}" class="circle responsive-img"></div> {% for category in page.categories %}<h5><a href="{{ category | prepend: '/categories/' | prepend: site.baseurl }}">{{ category }}</a>{% unless forloop.last%}, {% endunless %}{% endfor%} - {{ page.date | date: "%m/%d/%Y" }}</h5>
</div>
    {% for tag in page.tags %}
    {% if tag != site.sticky %}
    <a href="{{ tag | prepend: '/tags/' | prepend: site.baseurl }}"><div class="chip waves-effect waves-red">{{ tag }}</div></a>
    {% endif %}
    {% endfor %}
</div>
<div class="row">
{{ content }}
</div>

I don't know how can I fix this problem.

**Comment by neilernst** [comment]
I had an older version of MM where the `_config.yml` key was `plugins_dir`. Renaming that to `plugins` fixed the issue.
